pkg/types/types.go

```diff
@@ -0,0 +1,140 @@
+package types
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/google/uuid"
+	"go.hollow.sh/toolbox/events"
+	"golang.org/x/exp/slices"
+)
+
+// TODO: move into a shared package
+
+// ConditionKind defines model for ConditionKind.
+type ConditionKind string
+
+const (
+	FirmwareInstallOutofband ConditionKind = "firmwareInstallOutofband"
+	InventoryOutofband       ConditionKind = "inventoryOutofband"
+	ServerResourceType       string        = "servers"
+
+	ConditionEventType events.EventType = "condition"
+)
+
+// ConditionState defines model for ConditionState.
+type ConditionState string
+
+// Defines values for ConditionState.
+const (
+	Pending   ConditionState = "pending"
+	Active    ConditionState = "active"
+	Failed    ConditionState = "failed"
+	Succeeded ConditionState = "succeeded"
+)
+
+// ConditionStates returns available condition states.
+func ConditionStates() []ConditionState {
+	return []ConditionState{
+		Active,
+		Pending,
+		Failed,
+		Succeeded,
+	}
+}
+
+// Transition valid returns a bool value if the state transition is allowed.
+func (current ConditionState) TransitionValid(next ConditionState) bool {
+	switch {
+	// Pending state can stay in Pending or transition to Active or Failed or Succeeded
+	case current == Pending && slices.Contains([]ConditionState{Pending, Active, Failed, Succeeded}, next):
+		return true
+	// Active state can stay in Active or transition to Failed or Succeeded
+	case current == Active && slices.Contains([]ConditionState{Active, Failed, Succeeded}, next):
+		return true
+	default:
+		return false
+	}
+}
+
+// ConditionStateValid validates the ConditionState.
+func ConditionStateValid(s ConditionState) bool {
+	return slices.Contains(ConditionStates(), s)
+}
+
+// ConditionStateFinalized returns true when the state is finalized.
+func ConditionStateFinalized(s ConditionState) bool {
+	return slices.Contains([]ConditionState{Failed, Succeeded}, s)
+}
+
+// Parameters is an interface for Condition Parameter types
+type Parameters interface {
+	Validate() error
+}
+
+// ConditionDefinition holds the default parameters for a Condition.
+type ConditionDefinition struct {
+	Kind                  ConditionKind `mapstructure:"kind"`
+	Exclusive             bool          `mapstructure:"exclusive"`
+	FailOnCheckpointError bool          `mapstructure:"failOnCheckpointError"`
+	// TODO: we might want to be able to pass some parameters to the next condition.
+	OnSuccessCondition ConditionKind `mapstructure:"onSuccessCondition"`
+}
+
+// ConditionDefinitions is the list of conditions with helper methods.
+type ConditionDefinitions []*ConditionDefinition
+
+func (c ConditionDefinitions) FindByKind(k ConditionKind) *ConditionDefinition {
+	for _, e := range c {
+		if e.Kind == k {
+			return e
+		}
+	}
+
+	return nil
+}
+
+// Condition defines model for Condition.
+type Condition struct {
```

Comment:
As we discussed earlier, we probably want at least two other fields: `ID` (of some kind, preferably string or UUID), `structureVersion` (so we can control upgrades and reading historical data; this probably does not need to be an exported field). Finally we might want to use an exported `Version` or `Revision` field to provide for the locking/updating semantics you reference in the K8s API doc under `concurrency control`.

Reply:
> we probably want at least two other fields: ID (of some kind, preferably string or UUID), `structureVersion`

[d278100](https://github.com/metal-toolbox/conditionorc/pull/1/commits/d27810004464f02ac283fe31b783655d6e054324) adds an ID field, further work on will be followed up in https://github.com/metal-toolbox/conditionorc/issues/3

[e2db03f](https://github.com/metal-toolbox/conditionorc/pull/1/commits/e2db03ffbdaec72525bd5120aacb6115941aeba2) adds the `Version` field - let me know if this is too ambiguous to begin with.

> Finally we might want to use an exported Version or Revision field to provide for the locking/updating semantics you reference in the K8s API doc under concurrency control.

The current [`ResourceVersion`](https://github.com/metal-toolbox/conditionorc/blob/e2db03ffbdaec72525bd5120aacb6115941aeba2/pkg/types/types.go#L133)  covers this ^, unless I'm missing something.